File: test/jython/dbtest/hbase/test_query.py
#
# Tests out usage of the general HBase query interface
#

from jyunit.util import *

import java.lang

import db.hbase
from org.apache.hadoop.hbase.client import HTablePool
from meetup.beeno import EntityService, Query, Criteria, HBaseException
from meetup.beeno.util import HUtil
from meetup.beeno import TestEntities
from meetup.beeno.mapping import EntityMetadata
from dbtest.hbase import HBaseContext

hc = HBaseContext()

def setup():
    hc.setUp()
    HUtil.setPool( HTablePool( hc.conf, 5 ) )
    # create a dummy HBase table for testing
    import db.hbase
    admin = db.hbase.Admin(hc.conf)

    if not admin.exists("test_indexed"):
        admin.create("test_indexed", {"props:": {}})	
    if not admin.exists("test_indexed-by_intcol"):
        admin.create("test_indexed-by_intcol", {"props:": {}, "__idx__:": {}})
    if not admin.exists("test_indexed-by_stringcol"):
        admin.create("test_indexed-by_stringcol", {"props:": {}, "__idx__:": {}})

    srv = EntityService(TestEntities.IndexedEntity)
    now = java.lang.System.currentTimeMillis()

    srv.save( TestEntities.IndexedEntity("e1", "duck", 1, now - 100) )
    srv.save( TestEntities.IndexedEntity("e2", "duck", 2, now - 80) )
    srv.save( TestEntities.IndexedEntity("e3", "duck", 2, now - 60) )
    srv.save( TestEntities.IndexedEntity("e4", "goose", 2, now - 40) )


def teardown():
    try:
        pass
        # clean up the dummy table
        import db.hbase
        admin = db.hbase.Admin(hc.conf)

        if admin.exists("test_indexed"):
            admin.drop("test_indexed")
        if admin.exists("test_indexed-by_intcol"):
            admin.drop("test_indexed-by_intcol")
        if admin.exists("test_indexed-by_stringcol"):
            admin.drop("test_indexed-by_stringcol")
    finally:
        hc.tearDown()
        # hack to give server time to shutdown
        java.lang.Thread.sleep(10000)


def query_by_string():
    srv = EntityService(TestEntities.IndexedEntity)
    # test indexing of a value with multiple entries
    q = srv.query()
    q.using( Criteria.eq( "stringProperty", java.lang.String('duck') ) )
    matches = q.execute()

    assertEquals( len(matches), 3 )
    assertEquals( matches[0].getId(), 'e1' )
    assertEquals( matches[0].getStringProperty(), 'duck' )
    assertEquals( matches[1].getId(), 'e2' )
    assertEquals( matches[1].getStringProperty(), 'duck' )
    assertEquals( matches[2].getId(), 'e3' )
    assertEquals( matches[2].getStringProperty(), 'duck' )

    q = srv.query()
    q.using( Criteria.eq( "stringProperty", java.lang.String('goose') ) )
    matches = q.execute()
    assertEquals( len(matches), 1 )
    assertEquals( matches[0].getId(), 'e4' )
    assertEquals( matches[0].getStringProperty(), 'goose' )


def query_by_int():
    srv = EntityService(TestEntities.IndexedEntity)
    # test indexing of integer values
    q = srv.query()
    q.using( Criteria.eq( "intKey", java.lang.Integer(2) ) )
    matches = q.execute()
    print matches
    assertEquals( len(matches), 3 )
    assertEquals( matches[0].getId(), 'e4', "Indexed entries should be in reverse timestamp order" )
    assertEquals( matches[0].getIntKey(), 2 )
    assertEquals( matches[1].getId(), 'e3', "Indexed entries should be in reverse timestamp order" )
    assertEquals( matches[1].getIntKey(), 2 )
    assertEquals( matches[2].getId(), 'e2', "Indexed entries should be in reverse timestamp order" )
    assertEquals( matches[2].getIntKey(), 2 )

    q = srv.query()
    q.using( Criteria.eq( "intKey", java.lang.Integer(1) ) )
    matches = q.execute()
    assertEquals( len(matches), 1 )
    assertEquals( matches[0].getId(), 'e1' )
    assertEquals( matches[0].getIntKey(), 1 )


def run_test():
    query_by_string()
    query_by_int()


if __name__ == '__main__':
    try:
        setup()
        run_test()
    finally:
        teardown()

File: src/test/java/meetup/beeno/TestQuery.java
package meetup.beeno;

import meetup.beeno.util.HUtil;
import org.apache.hadoop.hbase.HBaseTestingUtility;
import org.apache.hadoop.hbase.client.HTablePool;
import org.apache.hadoop.hbase.util.Bytes;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import java.util.List;

import static org.junit.Assert.assertEquals;

/**
 */
public class TestQuery {
  private static HBaseTestingUtility UTIL;

  @BeforeClass
  public static void setupBeforeClass() throws Exception {
    UTIL = new HBaseTestingUtility();
    UTIL.startMiniCluster();
    HUtil.setPool(new HTablePool(UTIL.getConfiguration(), 5));
    // create a dummy HBase table for testing
    UTIL.createTable(Bytes.toBytes("test_indexed"), Bytes.toBytes("props"));
    UTIL.createTable(Bytes.toBytes("test_indexed-by_intcol"), Bytes.toByteArrays(new String[]{"props", "__idx__"}));
    UTIL.createTable(Bytes.toBytes("test_indexed-by_stringcol"), Bytes.toByteArrays(new String[]{"props", "__idx__"}));

    EntityService<Entities.IndexedEntity> srv = new EntityService(Entities.IndexedEntity.class);
    long now = java.lang.System.currentTimeMillis();

    srv.save( new Entities.IndexedEntity("e1", "duck", 1, now - 100) );
    srv.save( new Entities.IndexedEntity("e2", "duck", 2, now - 80) );
    srv.save( new Entities.IndexedEntity("e3", "duck", 2, now - 60) );
    srv.save( new Entities.IndexedEntity("e4", "goose", 2, now - 40) );
  }

  @AfterClass
  public static void tearDownAfterClass() throws Exception {
    UTIL.shutdownMiniCluster();
  }

  @Test
  public void queryByString() throws Exception {
    EntityService<Entities.IndexedEntity> srv = new EntityService(Entities.IndexedEntity.class);
    // test indexing of a value with multiple entries
    Query<Entities.IndexedEntity> q = srv.query();
    q.using( Criteria.eq( "stringProperty", "duck" ) );
    List<Entities.IndexedEntity> matches = q.execute();

    assertEquals( 3, matches.size() );
    assertEquals("e1", matches.get(0).getId());
    assertEquals("duck", matches.get(0).getStringProperty());
    assertEquals("e2", matches.get(1).getId());
    assertEquals("duck", matches.get(1).getStringProperty());
    assertEquals("e3", matches.get(2).getId());
    assertEquals("duck",  matches.get(2).getStringProperty());

    q = srv.query();
    q.using( Criteria.eq( "stringProperty", "goose" ) );
    matches = q.execute();
    assertEquals( 1, matches.size() );
    assertEquals("e4", matches.get(0).getId());
    assertEquals("goose", matches.get(0).getStringProperty());
  }

  @Test
  public void queryByInt() throws Exception {
    EntityService<Entities.IndexedEntity> srv = new EntityService(Entities.IndexedEntity.class);
    // test indexing of integer values
    Query<Entities.IndexedEntity> q = srv.query();
    q.using( Criteria.eq( "intKey", 2 ) );
    List<Entities.IndexedEntity> matches = q.execute();

    assertEquals( 3, matches.size() );
    assertEquals("Indexed entries should be in reverse timestamp order", "e4", matches.get(0).getId());
    assertEquals(new Integer(2), matches.get(0).getIntKey());
    assertEquals("Indexed entries should be in reverse timestamp order", "e3",  matches.get(1).getId());
    assertEquals(new Integer(2), matches.get(1).getIntKey());
    assertEquals("Indexed entries should be in reverse timestamp order", "e2",  matches.get(2).getId());
    assertEquals(new Integer(2), matches.get(2).getIntKey());

    q = srv.query();
    q.using( Criteria.eq( "intKey", 1 ) );
    matches = q.execute();
    assertEquals( 1, matches.size() );
    assertEquals("e1", matches.get(0).getId());
    assertEquals(new Integer(1), matches.get(0).getIntKey());
  }
}
